pkg/analyze/event.go

```diff
@@ -0,0 +1,196 @@
+package analyzer
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	troubleshootv1beta2 "github.com/replicatedhq/troubleshoot/pkg/apis/troubleshoot/v1beta2"
+	"github.com/replicatedhq/troubleshoot/pkg/constants"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+)
+
+type AnalyzeEvent struct {
+	analyzer *troubleshootv1beta2.EventAnalyze
+}
+
+type eventFilter struct {
+	kind     string
+	reason   string
+	msgRegex string
+}
+
+func (a *AnalyzeEvent) Title() string {
+	if a.analyzer.CheckName != "" {
+		return a.analyzer.CheckName
+	}
+	if a.analyzer.CollectorName != "" {
+		return a.analyzer.CollectorName
+	}
+	return "Event"
+}
+
+func (a *AnalyzeEvent) IsExcluded() (bool, error) {
+	return isExcluded(a.analyzer.Exclude)
+}
+
+func (a *AnalyzeEvent) Analyze(getFile getCollectedFileContents, findFiles getChildCollectedFileContents) ([]*AnalyzeResult, error) {
+	// read collected events based on namespace
+	namespace := getNamespace(a.analyzer.Namespace)
+	fullPath := path.Join(constants.CLUSTER_RESOURCES_DIR, constants.CLUSTER_RESOURCES_EVENTS, namespace)
+	fullPath = fmt.Sprintf("%s.json", fullPath)
+	fileContent, err := getFile(fullPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read collected events for namespace: %s", namespace)
+	}
+
+	events, err := convertToEventList(fileContent)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read collected events")
+	}
+
+	// filter if there's single event matched with the given criteria
+	// match: Reason && Kind (optional) && MessageRegex (optional)
+	// e.g. Reason: Unhealthy. Kind: Pod. Message: Readiness probe failed:...
+	event := getEvent(events, eventFilter{
+		kind:     a.analyzer.Kind,
+		reason:   a.analyzer.Reason,
+		msgRegex: a.analyzer.RegexPattern,
+	})
+
+	return analyzeEventResult(event, a.analyzer.Outcomes, a.Title())
+
+}
+
+func getNamespace(namespace string) string {
+	if namespace == "" {
+		return corev1.NamespaceDefault
+	}
+	return namespace
+}
+
+func convertToEventList(data []byte) (*corev1.EventList, error) {
+	var eventList corev1.EventList
+	err := json.Unmarshal(data, &eventList)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert []byte to corev1.EventList: %w", err)
+	}
+	return &eventList, nil
+}
+
+func getEvent(events *corev1.EventList, filter eventFilter) *corev1.Event {
+	for _, event := range events.Items {
+		if !matchReason(event.Reason, filter.reason) {
```

Comment:
It looks like that `Reason` is must have field. It filter reason first. 
One question, is that possible that user just need to add `regex` without reason? Like Diamon's example
https://github.com/replicatedhq/troubleshoot/issues/911#issue-1498626521

Reply:
I am thinking of 2 approaches:

1. User must know the specific `Reason` and the `regex` is only for `Message`. We can document this https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/events/event.go
2. User only need to provide `regex` and the `regex` will be matched on both `Reason` and `Message`. I am concern this will have performance impact as the current implementation we are iterating over all Events.

Please let me know your thought and I will update the code accordingly.